// ngx-extended-pdf-viewer/esm5/lib/pinch-on-mobile-support.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pinch-on-mobile-support.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var PinchOnMobileSupport = /** @class */ (function () {
    function PinchOnMobileSupport(_zone) {
        this._zone = _zone;
        this.startX = 0;
        this.startY = 0;
        this.initialPinchDistance = 0;
        this.pinchScale = 1;
        if (this.isMobile()) {
            this.initializePinchZoom();
        }
    }
    /**
     * @private
     * @return {?}
     */
    PinchOnMobileSupport.prototype.isMobile = /**
     * @private
     * @return {?}
     */
    function () {
        return 'ontouchstart' in document.documentElement;
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PinchOnMobileSupport.prototype.onViewerTouchStart = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.initialPinchDistance = 0;
        if (event.touches.length === 2) {
            /** @type {?} */
            var rect = this.container.getBoundingClientRect();
            // + this.container.scrollTop
            if (event.touches[0].pageX >= rect.left && event.touches[0].pageX <= rect.right) {
                if (event.touches[0].pageY >= rect.top /* && event.touches[0].pageY <= rect.bottom */) {
                    if (event.touches[1].pageX >= rect.left && event.touches[1].pageX <= rect.right) {
                        if (event.touches[1].pageY >= rect.top /* && event.touches[1].pageY <= rect.bottom  */) {
                            this.startX = (event.touches[0].pageX + event.touches[1].pageX) / 2;
                            this.startY = (event.touches[0].pageY + event.touches[1].pageY) / 2;
                            this.initialPinchDistance = Math.hypot(event.touches[1].pageX - event.touches[0].pageX, event.touches[1].pageY - event.touches[0].pageY);
                            event.preventDefault();
                            event.stopPropagation();
                        }
                    }
                }
            }
        }
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PinchOnMobileSupport.prototype.onViewerTouchMove = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.initialPinchDistance <= 0 || event.touches.length !== 2) {
            return;
        }
        if (this.pinchScale !== 1) {
            event.preventDefault();
        }
        /** @type {?} */
        var pinchDistance = Math.hypot(event.touches[1].pageX - event.touches[0].pageX, event.touches[1].pageY - event.touches[0].pageY);
        /** @type {?} */
        var originX = this.startX + this.container.scrollLeft;
        /** @type {?} */
        var originY = this.startY + this.container.scrollTop;
        this.pinchScale = pinchDistance / this.initialPinchDistance;
        this.viewer.style.transform = "scale(" + this.pinchScale + ")";
        this.viewer.style.transformOrigin = originX + "px " + originY + "px";
        event.preventDefault();
        event.stopPropagation();
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PinchOnMobileSupport.prototype.onViewerTouchEnd = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var PDFViewerApplication = ((/** @type {?} */ (window))).PDFViewerApplication;
        if (this.initialPinchDistance <= 0) {
            return;
        }
        this.viewer.style.transform = "none";
        this.viewer.style.transformOrigin = "unset";
        PDFViewerApplication.pdfViewer.currentScale *= this.pinchScale;
        /** @type {?} */
        var rect = this.container.getBoundingClientRect();
        /** @type {?} */
        var dx = this.startX - rect.left;
        /** @type {?} */
        var dy = this.startY - rect.top;
        this.container.scrollLeft += dx * (this.pinchScale - 1);
        this.container.scrollTop += dy * (this.pinchScale - 1);
        this.resetPinchZoomParams();
        event.preventDefault();
        event.stopPropagation();
    };
    /**
     * @private
     * @return {?}
     */
    PinchOnMobileSupport.prototype.resetPinchZoomParams = /**
     * @private
     * @return {?}
     */
    function () {
        this.startX = this.startY = this.initialPinchDistance = 0;
        this.pinchScale = 1;
    };
    /**
     * @return {?}
     */
    PinchOnMobileSupport.prototype.initializePinchZoom = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.viewer = document.getElementById('viewer');
        this.container = (/** @type {?} */ (document.getElementById('viewerContainer')));
        this._zone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            document.addEventListener('touchstart', _this.onViewerTouchStart.bind(_this));
            document.addEventListener('touchmove', _this.onViewerTouchMove.bind(_this), { passive: false });
            document.addEventListener('touchend', _this.onViewerTouchEnd.bind(_this));
        }));
    };
    /**
     * @return {?}
     */
    PinchOnMobileSupport.prototype.destroyPinchZoom = /**
     * @return {?}
     */
    function () {
        if (this.isMobile()) {
            document.removeEventListener('touchstart', this.onViewerTouchStart);
            document.removeEventListener('touchmove', this.onViewerTouchMove);
            document.removeEventListener('touchend', this.onViewerTouchEnd);
        }
    };
    return PinchOnMobileSupport;
}());
export { PinchOnMobileSupport };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.viewer;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.container;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.startX;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.startY;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.initialPinchDistance;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype.pinchScale;
    /**
     * @type {?}
     * @private
     */
    PinchOnMobileSupport.prototype._zone;
}
